Support FMA operation

thvnx opened the pull request:
This commit adds the floating-point fused multiply-add operation (FMA) to the Pervasives module.

The following changes are made:
- _configure_: check `math.h` for the C99 `fma()` operation.
- fma declarations in _pervasives.ml[i]_ (stdlib and otherlibs/threads).
- C `fma()` call or emulation in _byterun/floats.c_.
- Dedicated tests.

I am a regular ocaml follower, as well as an intensive FMA user. I've done this work to enjoy both ocaml language and FMA operation. If useful to the community, could this contribution be integrated to the next ocaml release?

The test suite passes successfully.

thvnx commented:
When emulated (only activated if C99 functions aren't provided by math.h), the FMA doesn't resist to overflows or underflows. Do I have to mention it in documentation?

alainfrisch commented:
Can you confirm that this is intended to improve the precision of the calculation, not to make the combined operation faster (which wouldn't probably be the case because of the C function call)?

thvnx commented:
You are right. It's for accuracy purposes, and additionally, improves the programming of extended-precision-based numerical algorithms (with perhaps performance improvements in some such cases).

It's very usefull for me but maybe not so relevant for the ocaml developers community?

alainfrisch commented:
I've a biais towards numerical code, so I'll let other core maintainers comment on the relevance of including the function in the stdlib.  (The fact that it could at some point be implemented directly in the backend to benefit from CPU instructions would be an argument in favor of including the function.)

Anyway, if it is decided that the function deserves to be in the stdlib, it should probably go in the future Float module discussed elsewhere, not in Pervasives.

thvnx commented:
Thanks for your comments. I agree.

Where can I find the discussion about the Float module?

xavierleroy commented:
I agree with @alainfrisch that this would be a fine addition to the future `Float` module.  (Perhaps along with other C99 FP functions that we still miss.)

The fallback implementation is taken from authors I trust.  However it uses some C99 features (declaration in the middle of a block) that (some version of) MSVC doesn't support.

It should be easy to have `ocamlopt` turn calls to this function into FMA instructions provided by the processor, if they exist.  E.g. for PowerPC and for ARM64, possibly for ARM too.  For x86-64 FMA instructions were added to SSE only recently, so we have a portability issue.

alainfrisch commented:
Adding a Float module was discussed in #964, and this is now conditioned on #1010.

jhjourdan commented:
Is the fallback implementation correct if overflow occur? The paper which is cited explicitly exclude this case from its proofs.

I am not sure whether this applies here, but note that overflow could occur in intermediate computations but not in the correct final result. In that case, it is even not enough to add something like "undefined result in case of overflow" in the documentation, because then it is not clear what overflow mean.

We should also take care about infinites and NaN...

thvnx commented:
@jhjourdan, the fallback implementation isn't correct if overflow occurs. In that case, the result will be as if computations were done without FMA.

I'll look if a more robust FMA emulation exists (maybe [here](https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=sysdeps/ieee754/dbl-64/s_fma.c;hb=HEAD)).

EDIT: emulation linked above seems conform to IEEE standard.

thvnx commented:
Does anyone understand why AppVeyor build fails? I don't have any Windows environment to reproduce the actual issue, i.e. warnings from line 1768 to 1773.

nojb commented:
This is related to some recent PR merges and it is being worked on.  It should be resolved in the near future.

xavierleroy commented:
Concerning overflows, ISO C99 7.12.13.1 says "A range error may occur", without further details.  So, I'm not sure all `fma` implementations out there handle overflow in `x * y` correctly.

thvnx commented:
About the FMA emulation: the one provided by the GLIBC is conform to IEEE 754 requirements (since the version 2.24, see [here](https://sourceware.org/bugzilla/show_bug.cgi?id=13304)). Note that it uses the same algorithm which is currently implemented in this PR. I will soon adapt the fallback consequently.

jhjourdan commented:
Great !

thvnx commented:
I've updated the FMA emulation and added new tests. FMA fallback supports now overflow and almost all underflows (some double-rounding issues are very difficult to handle: 3 of 317 tests still fail).

damiendoligez commented:
Now we have merged #1010 and #1638, so we have a `Float` module. @thvnx could you update this PR?

thvnx commented:
Great! I'll do it ASAP.

thvnx committed:
Support FMA operation

This commit adds a fused multiply-add operation to the Float module.

The following changes are made:
- configure: check math.h for the C99 fma() operation.
- fma declarations in float.ml[i] (stdlib/).
- C fma() call or emulation in runtime/floats.c.
- dedicated tests in testsuite/tests/fma.

thvnx commented:
I moved the fma operation in the Float module.

But, I still have a few things to do:

- [x] 1. Make tests working again. testsuite/HACKING.adoc is incomplete, directives for 'Creating a new test' are missing. I do not understand how to add new tests well for now.
- [ ] 2. In case of fma emulation only (when C99 functions are not available), some double-rounding issues are very difficult to handle: 3 of 317 tests still fail.
- [ ] 3. For the tests, I've taken all the numerical values from the glibc ones, do I need to be more specific in testsuite/tests/fma/fma.ml (with a link to the original file for example)?

damiendoligez commented:
> `testsuite/HACKING.adoc` is incomplete, directives for 'Creating a new test' are missing. I do not understand how to add new tests well for now.

cc @shindere

shindere commented:
Well I have to say I never paid attention to `testsuite/HACKING.adoc`',
sorry about that. I just looked at it and it seems it's a work in
progress.

@thvnx did you try to look at existing tests to see how they work?

If you are able to describe the kind of test you would like to add, I
can describe how you should do.

thvnx commented:
@shindere, yes, I have a .ml file with its .reference counterpart. I've added an ocamltests file which contains the file name of my .ml test. When I run 'make tests', I obtain the following output:

```
List of directories returning no results:
    tests/fma
```

shindere commented:
You need to add an ocamltest block at the beginning of your file. For
instance

```ocaml
(* TEST *)
```

may be enough. You can run that test directly by doing, from the
directory in question:

```bash
../../../ocamltest/ocamltest test.ml
```

You can also run the testsuite just on this directory by running the
following command from the testsuite directory:

```bash
make one DIR=tests/fma
```

thvnx commented:
Thanks, It's done!

thvnx commented:
It's almost done, I still have to fix the FMA emulation fallback to pass all the remaining failing tests.

thvnx commented:
I did not succeed in solving the three failed tests for the hard-to-tackle double-rounding issue (in the case of fallback emulation). So, I just disabled them. It works well when FMA hardware is provided but not fully IEEE compliant in emulation mode. OK anyway?

alainfrisch commented on stdlib/float.mli:
```diff
@@ -57,6 +57,13 @@ external mul : float -> float -> float = "%mulfloat"
+external fma : float -> float -> float -> float =
+  "caml_fma_float" "caml_fma" [@@unboxed] [@@noalloc]
+(** A fused multiply-add [fma x y z] computes [x * y + z] with a
+   single rounding. Deprecated if no hardware FMA support is provided,
```
I'm not sure "deprecated" is relevant here.  Something like "If no hardware support is provided, ...".  But can this also be caused by a lack of libc support?  And if the libc provides the emulation itself without hardware support, is it fully accurate?

thvnx replied:
It depends on the libc. Glibc is IEEE compliant, I don't know about the others. The fallback I wrote is very close to the one of the glibc (but they can switch rounding modes and other things that I can't do here).

alainfrisch commented:
>  OK anyway?

Ok for me, but perhaps @xavierleroy or @Chris00 would like to comment on it.

thvnx committed:
fixup! Support FMA operation

alainfrisch commented on stdlib/float.mli:
```diff
@@ -60,7 +60,7 @@ external div : float -> float -> float = "%divfloat"
   "caml_fma_float" "caml_fma" [@@unboxed] [@@noalloc]
 (** A fused multiply-add [fma x y z] computes [x * y + z] with a
-   single rounding. Deprecated if no hardware FMA support is provided,
+   single rounding. If no hardware support is provided,
```
Ok, so IIUC, the criterion is not so much that the hardware support is provided or not, but whether the libc provides a compliant version (emulated or hardware based) or not.  There can be cases where hardware support is available but emulation is still used (your emulation, or from the libc).  (Sorry to nitpick, but it's important to explain as precisely possible sources of differences in fp calculations across different machines.)

thvnx replied:
Yes, you're absolutely right! I'll fix it in the next commit.

thvnx committed:
fixup! Support FMA operation

alainfrisch commented on stdlib/float.mli:
```diff
@@ -60,9 +60,9 @@ external div : float -> float -> float = "%divfloat"
-   double-rounding issue may happen with software emulation fallback.
-   @since 4.09.0 *)
+   single rounding. Can be not fully IEEE compliant if a software
+   emulation is use in place of the hardware instruction. @since
```
Proposal:
````
[fma x y z] returns [x * y + z], with a best effort for computing this expression with a single rounding, using either hardware instructions (providing full IEEE compliance) or a software emulation.

@since 4.08
````

thvnx replied:
Fixed, thanks!

thvnx committed:
fixup! Support FMA operation

alainfrisch commented on configure:
```diff
@@ -1064,7 +1064,7 @@ fi
 # For the Pervasives module
 
 if sh ./hasgot2 -i math.h $mathlib expm1 log1p hypot copysign; then
```
You should include `fma` in the list of functions tested with hasgot2.

thvnx replied:
Done

alainfrisch commented on stdlib/float.mli:
```diff
@@ -57,6 +57,13 @@ external mul : float -> float -> float = "%mulfloat"
+(** [fma x y z] returns [x * y + z], with a best effort for computing
+    this expression with a single rounding, using either hardware
+    instructions (providing full IEEE compliance) or a software emulation.
+    @since 4.09.0 *)
```
1. 4.08
2. I did not think about it before, but what about saying a word about performance (with and without emulation) compared to the naive version?  People might think that using fma is a good idea for performance reason even when precision is not a big concern, and it's not obvious whether this is a good idea.  Is it true that with hardware support, fma would be fastest but with emulation from libc or yours, itwould be slower?

thvnx replied:
Add a note about performance.

alainfrisch commented on Changes:
```diff
@@ -126,6 +126,9 @@ Working version
   (Christophe Troestler, review by Alain Frish, Xavier Clerc and Daniel Bünzli)
 
+- GPR#1354: Add fma support to Float module.
+  (Laurent Thévenoux)
```
Please include people involved in the review.

thvnx replied:
Reviewers only or all participants?

alainfrisch replied:
Looking quickly at the discussion, I'd say : @jhjourdan @xavierleroy @alainfrisch

thvnx replied:
Thanks, done!

alainfrisch commented on testsuite/tests/fma/fma.ml:
```diff
@@ -0,0 +1,472 @@
+
+(* tests 254, 257--259, 264--265, 267--268, and 272--273 are disabled
+   because of unresolved double-rounding issue in FMA emulation
+   fallback *)
```
I'll let you decide, but perhaps, for those tests, you could specify either a list of possible results (the correct one, and the one obtained by the emulation), or do a check "modulo epsilon"; just to make sure that the function is not too crazy in those cases.

thvnx replied:
Fixed (with list of possible results), integration in progress, I'm waiting for mscv64 failures if any.

alainfrisch reviewed (approved):
Minor comments only.

I'm not competent to check the emulation code but @xavierleroy said he had enough trust, so I think this can be merged once comments are addressed unless someone else speaks up.

thvnx committed:
Take reviewer comments into account

thvnx committed:
fixup! Take reviewer comments into account

thvnx committed:
fixup! Take reviewer comments into account

alainfrisch commented on Changes:
```diff
@@ -126,6 +126,10 @@ Working version
   (Christophe Troestler, review by Alain Frish, Xavier Clerc and Daniel Bünzli)
 
+- GPR#1354: Add fma support to Float module.
+  (Laurent Thévenoux, review by Alain Frish, Jacques-Henri Jourdan,
```
```suggestion
  (Laurent Thévenoux, review by Alain Frisch, Jacques-Henri Jourdan,
```

thvnx replied:
Sorry!

alainfrisch committed:
Fix typo

Co-Authored-By: thvnx <laurent.thevenoux@gmail.com>

thvnx committed:
Improve test comments

alainfrisch commented:
Thanks @thvnx for your contribution and patience when dealing with incremental comments!

shindere commented:
This PR causes a failure of the fma test on Inria's CI (Cygwin32 machine).
The problem happens wiht and without flambda in native and bytecode mode.
There is a difference between the expected output and the obtained one:

```
@@ -252,17 +252,17 @@
 252 OK!
 253 OK!
 254 OK!
-255 OK!
+255 FAIL!	fma (0x1.fffp+0, 0x1.0000000000001p+0, -0x1.fffp+0) returned 0x1p-51 instead of 0x1.fffp-52.
 256 OK!
 257 OK!
-258 OK!
-259 OK!
-260 OK!
+258 FAIL!	fma (0x1.deadbeef2feedp+1023, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp+1) returned 0x1.098p-53 instead of 0x1.0989687bc9da4p-53.
+259 FAIL!	fma (0x1.deadbeef2feedp+900, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp-122) returned 0x1.098p-176 instead of 0x1.0989687bc9da4p-176.
+260 FAIL!	fma (0x1.fffffffffffffp+1023, 0x1.001p+0, -0x1.fffffffffffffp+1023) returned 0x1p+1012 instead of 0x1.fffffffffffffp+1011.
 261 OK!
 262 OK!
 263 OK!
-264 OK!
-265 OK!
+264 FAIL!	fma (0x1.deadbeef2feedp-495, 0x1.deadbeef2feedp-495, -0x1.bf86a5786a574p-989) returned 0x0.00000428p-1022 instead of 0x0.0000042625a1fp-1022.
+265 FAIL!	fma (0x1.deadbeef2feedp-503, 0x1.deadbeef2feedp-503, -0x1.bf86a5786a574p-1005) returned 0x0.000000000428p-1022 instead of 0x0.0000000004262p-1022.
 266 OK!
 267 OK!
 268 OK!
@@ -295,10 +295,10 @@
 295 OK!
 296 OK!
 297 OK!
-298 OK!
-299 OK!
-300 OK!
-301 OK!
+298 FAIL!	fma (0x1.fffffffffffffp-1, 0x1.fffffffffffffp-1, -0x1.ffffffffffffep-1) returned 0x0p+0 instead of 0x1p-106.
+299 FAIL!	fma (0x1.fffffffffffffp-1, -0x1.fffffffffffffp-1, 0x1.ffffffffffffep-1) returned 0x0p+0 instead of -0x1p-106.
+300 FAIL!	fma (-0x1.fffffffffffffp-1, 0x1.fffffffffffffp-1, 0x1.ffffffffffffep-1) returned 0x0p+0 instead of -0x1p-106.
+301 FAIL!	fma (-0x1.fffffffffffffp-1, -0x1.fffffffffffffp-1, -0x1.ffffffffffffep-1) returned 0x0p+0 instead of 0x1p-106.
 302 OK!
 303 OK!
 304 OK!
```

thvnx commented:
@shindere , I can easily fix it. Do I have to create a new pull request or I can just pull in this one?

alainfrisch commented:
No unfortunately, you need to create a new PR.  Is the problem a bad software emulation from that libc?

thvnx commented:
@alainfrisch, OK. Yes, I guess this is due to another software support for the fma on this platform (or something related to subnormals).

@shindere, I'll create a new PR with this patch: https://github.com/thvnx/ocaml/commit/06d120cd98d2c62955d09734cdb655ed0ce5b51b. Can you test it before someone merge it to trunk?

See #2120

shindere commented:
For several weeks, Inria's CI was broken in several ways, so that
the testsuite could not be run for the Cygwin 64 port.

Now that this has been fixed, it turns out that the FMA tests fail on
this platform, for both native and bytecode, with and without flambda.

More precisely, in all these cases the test's output differs from the
expected one:

```
@@ -255,15 +255,15 @@
 255 OK!
 256 OK!
 257 OK!
-258 OK!
-259 OK!
-260 OK!
-261 OK!
-262 OK!
+258 FAIL!	fma (0x1.deadbeef2feedp+1023, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp+1) returned 0x0p+0 instead of 0x1.0989687bc9da4p-53.
+259 FAIL!	fma (0x1.deadbeef2feedp+900, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp-122) returned 0x0p+0 instead of 0x1.0989687bc9da4p-176.
+260 FAIL!	fma (0x1.fffffffffffffp+1023, 0x1.001p+0, -0x1.fffffffffffffp+1023) returned infinity instead of 0x1.fffffffffffffp+1011.
+261 FAIL!	fma (-0x1.fffffffffffffp+1023, 0x1.fffffffffffffp+0, 0x1.fffffffffffffp+1023) returned -infinity instead of -0x1.ffffffffffffdp+1023.
+262 FAIL!	fma (0x1.fffffffffffffp+1023, 0x1p+1, -0x1.fffffffffffffp+1023) returned infinity instead of 0x1.fffffffffffffp+1023.
 263 OK!
-264 OK!
-265 OK!
-266 OK!
+264 FAIL!	fma (0x1.deadbeef2feedp-495, 0x1.deadbeef2feedp-495, -0x1.bf86a5786a574p-989) returned 0x0p+0 instead of 0x0.0000042625a1fp-1022.
+265 FAIL!	fma (0x1.deadbeef2feedp-503, 0x1.deadbeef2feedp-503, -0x1.bf86a5786a574p-1005) returned 0x0p+0 instead of 0x0.0000000004262p-1022.
+266 FAIL!	fma (0x1p-537, 0x1p-538, 0x0.0000000000001p-1022) returned 0x0.0000000000001p-1022 instead of 0x0.0000000000002p-1022.
 267 OK!
 268 OK!
 269 OK!
@@ -275,8 +275,8 @@
 275 OK!
 276 OK!
 277 OK!
-278 OK!
-279 OK!
+278 FAIL!	fma (0x0.0000000000001p-1022, 0x1p-1, 0x0.fffffffffffffp-1022) returned 0x0.fffffffffffffp-1022 instead of 0x1p-1022.
+279 FAIL!	fma (-0x0.0000000000001p-1022, 0x1p-1, -0x0.fffffffffffffp-1022) returned -0x0.fffffffffffffp-1022 instead of -0x1p-1022.
 280 OK!
 281 OK!
 282 OK!

```

And configure says:

```
checking for fma... yes
```

Any idea what may go wrong here?

dra27 commented:
@shindere - see #2177

shindere commented:
David Allsopp (2019/03/01 00:52 -0800):
> @shindere - see #2177

Yes sorry, indeed! Actually I think that one was waiting for you to take
a comment I did into account or just react to it saying you won't take
it into account ;-)
